designer/server/src/main/scala/pl/touk/nussknacker/ui/component/ComponentsUsageHelper.scala

```diff
@@ -15,21 +17,39 @@ object ComponentsUsageHelper {
   }
 
   def computeComponentsUsage(componentIdProvider: ComponentIdProvider,
-                             processesDetails: List[BaseProcessDetails[ScenarioComponentsUsages]]): Map[ComponentId, List[(BaseProcessDetails[Unit], List[NodeId])]] = {
-
-    def toComponentIdUsages(processDetails: BaseProcessDetails[ScenarioComponentsUsages]): List[(ComponentId, (BaseProcessDetails[Unit], List[NodeId]))] = {
-      val componentsUsages: Map[ComponentIdParts, List[NodeId]] = processDetails.json.value
-      componentsUsages.toList.map { case (ComponentIdParts(componentName, componentType), nodeIds) =>
-        val componentId = componentIdProvider.createComponentId(processDetails.processingType, componentName, componentType)
-        componentId -> (processDetails.mapProcess(_ => ()), nodeIds)
-      }
+                             processesDetails: List[BaseProcessDetails[ScenarioComponentsUsages]]): Map[ComponentId, List[(BaseProcessDetails[_], List[NodeMetadata])]] = {
+    def flattenUsages(processesDetails: List[BaseProcessDetails[ScenarioComponentsUsages]]) = for {
+      processDetails <- processesDetails
+      componentIdNodesPair <- processDetails.json.value.toList
+      (ComponentIdParts(componentName, componentType), nodeIds) = componentIdNodesPair
+      componentId = componentIdProvider.createComponentId(processDetails.processingType, componentName, componentType)
+      nodeId <- nodeIds
+      nodeMetadata = if (processDetails.isSubprocess) FragmentNodeMetadata(processDetails.name, nodeId) else ScenarioNodeMetadata(nodeId)
```

Comment:
Something is not yes here :) Why you cheating and putting scenario name into place where should be nodeId (fragmentNodeId)? IMO we would have `nodeId: String` after this step. You should rewrite `String` into `NodeMetadata` in next step when you resolve usages of fragments. OFC here:
```
      case usageOfOtherComponentType =>
        List(usageOfOtherComponentType)
```
will be necessary to replace nodeId by `ScenarioNodeMetadata` as well.

Other option I could see is to have more generic purpose structure - `NodeIdPath` which at the beginning will be one element list, and during resolving of fragment usages become `fragmentNodeId :: nodeId :: Nil`. On FE would be some logic that if the path has > 1 element1, then has other color.  Thanks to this approach it will be easier to migrate into nesting of fragments in the future

Reply:
> IMO we would have nodeId: String after this step. You should rewrite String into...

And indeed it is rewritten here:

```
val scenarioUsagesWithResolvedFragments = scenariosComponentUsagesFlatten.flatMap {
   case fragmentUsage@ScenarioComponentsUsage(_, ComponentType.Fragments, Some(fragmentName), processDetails, ScenarioNodeMetadata(fragmentNodeId)) =>
    fragmentUsage :: fragmentsComponentUsagesFlattenMap.get(fragmentName).toList.flatten.collect {
       case u@ScenarioComponentsUsage(componentId, componentType, componentName, _, FragmentNodeMetadata(_, nodeId)) =>
        u.copy(processDetails = processDetails, nodeMetadata = FragmentNodeMetadata(fragmentNodeId, nodeId))
    }
```

And I'm cheating here because in previous approach I had `NodeMetadata` structure with `Option` field for its fragment node id. With that solution I set fragment node id to `None` and rewritten it to `Some(...)` in the code above.